core/embed/extmod/modtrezorio/ff.c

```diff
@@ -3581,24 +3597,40 @@ FRESULT f_mkfs (
 			} else {
 				st_dword(buf + 0, (fsty == FS_FAT12) ? 0xFFFFF8 : 0xFFFFFFF8);	/* FAT[0] and FAT[1] */
 			}
-			nsect = sz_fat;		/* Number of FAT sectors */
+			nsect0 = nsect = sz_fat;		/* Number of FAT sectors */
 			do {	/* Fill FAT sectors */
 				n = (nsect > sz_buf) ? sz_buf : nsect;
 				if (disk_write(pdrv, buf, sect, (UINT)n) != RES_OK) LEAVE_MKFS(FR_DISK_ERR);
 				memset(buf, 0, ss);	/* Rest of FAT all are cleared */
 				sect += n; nsect -= n;
+
+				if (progress_callback) {
+					progress_callback(PROGRESS_SECTION(progress_before_initialize_fat_area, progress_initialize_fat_area_section, (1 - nsect / (float)nsect0)));
+				}
 			} while (nsect);
 		}
 
+		if (progress_callback) {
+			progress_callback(progress_before_initialize_root_directory);
+		}
+
 		/* Initialize root directory (fill with zero) */
-		nsect = (fsty == FS_FAT32) ? pau : sz_dir;	/* Number of root directory sectors */
+		nsect0 = nsect = (fsty == FS_FAT32) ? pau : sz_dir;	/* Number of root directory sectors */
 		do {
 			n = (nsect > sz_buf) ? sz_buf : nsect;
 			if (disk_write(pdrv, buf, sect, (UINT)n) != RES_OK) LEAVE_MKFS(FR_DISK_ERR);
 			sect += n; nsect -= n;
+
+			if (progress_callback) {
+				progress_callback(PROGRESS_SECTION(progress_before_initialize_root_directory, progress_initialize_root_directory_section, (1 - nsect / (float)nsect0)));
```

Comment:
The consequence of calling the progress callback on each iteration here (and above too) is that its actually called a lot. Each progress render takes ~ 10 ms, so there is actually quite difference in formatting time when showing the progress bar or not - almost 10 seconds on my 32GB sd card. 

I think we need to introduce here some throttling of the progress calls, i.e only call it lets say 20 times per the whole formatting

Reply:
https://github.com/trezor/trezor-firmware/pull/3700/commits/db422af0ce290aaabccd4777e44e7fba20fe8d1b#diff-65fb8326ce9d38d28ea6bceda8b732569da461b8b8f1d729f29c5538f1014fa3R3357